hamilton/function_modifiers/recursive.py

```diff
@@ -600,3 +603,292 @@ def required_config(self) -> Optional[List[str]]:
+        # need to access it from somewhere outside
+        self.upstream_dependency = pd.DataFrame
+
+    # TODO: This is duplicate from extract columns, probably can be combined
+    # TODO: Add async functionality
+    def _create_column_nodes(
+        self, inject_parameter: str, params: Dict[str, Type[Type]]
+    ) -> List[node.Node]:
+        """Helper function to extract columns for the user."""
+        output_type = params[inject_parameter]
+        series_type = registry.get_column_type_from_df_type(output_type)
+
+        out = []
+        for column in self.initial_schema:
+
+            def extractor_fn(
+                column_to_extract: str = column, **kwargs
+            ) -> Any:  # avoiding problems with closures
+                df = kwargs[inject_parameter]
+                if column_to_extract not in df:
+                    raise base.InvalidDecoratorException(
+                        f"No such column: {column_to_extract} produced by {inject_parameter}. "
+                        f"It only produced {str(df.columns)}"
+                    )
+                return registry.get_column(df, column_to_extract)
+
+            out.append(
+                node.Node(
+                    name=column,
+                    typ=series_type,
+                    callabl=extractor_fn,
+                    input_types={inject_parameter: output_type},
+                )
+            )
+
+        return out
+
+    def _get_inital_nodes(
+        self, fn: Callable, params: Dict[str, Type[Type]]
+    ) -> Tuple[str, Collection[node.Node]]:
+        """Selects the correct dataframe and optionally extracts out columns."""
+        initial_nodes = []
+        if self.dataframe_subdag_param is not None:
+            inject_parameter = self.dataframe_subdag_param
+        else:
+            # If we don't have a specified dataframe we assume it's the first argument
+            sig = inspect.signature(fn)
+            inject_parameter = list(sig.parameters.values())[0].name
+            initial_nodes.extend(
+                self._create_column_nodes(inject_parameter=inject_parameter, params=params)
+            )
+
+        if inject_parameter not in params:
+            raise base.InvalidDecoratorException(
+                f"Function: {fn.__name__} has a first parameter that is not a dependency. "
+                f"@with_columns requires the parameter names to match the function parameters. "
+                f"Thus it might not be compatible with some other decorators"
+            )
+
+        return inject_parameter, initial_nodes
+
+    def _create_merge_node(self, upstream_node: str, node_name: str) -> node.Node:
+        "Node that adds to / overrides columns for the original dataframe based on selected output."
+
+        def new_callable(**kwargs) -> Any:
+            df = kwargs[upstream_node]
+            for column in self.select:
+                df[column] = kwargs[column]
```

Comment:
Hmm, we're mutating this... We should be careful as that breaks assumptions. 

The nice thing about pandas is that the dataframes shouldn't mess with memory if you create new ones, the series are what takes up the memory. So we should be able to continually modify it by creating new ones (I think...)

Reply:
Might be worth using `assign` or something like that... Looking here (you mentioned similar things in your comments on the code): https://stackoverflow.com/questions/72291290/how-to-create-new-column-dynamically-in-pandas-like-we-do-in-pyspark-withcolumn

Reply:
Looking at the way [Narwhals implements](https://github.com/narwhals-dev/narwhals/blob/5c3db5b149cbf64b97de49bc4ad726f019f9f578/narwhals/_pandas_like/dataframe.py#L388) the Polars `.with_columns()` for other backends can also be informative.

Reply:
I looked a bit deeper:

1. `assign` does a copy of the data frame. (It should be shallow unless the flag is set to `copy="deep"`. From [this thread](https://stackoverflow.com/questions/48177914/why-use-pandas-assign-rather-than-simply-initialize-new-column) my impression is that assign was created for chaining operations together and leaving the original data frame intact.
2. Narwhals uses `concat(dfs, axis=1, **extra_kwargs)` under the hood and this does have a warning: "[concat()](https://pandas.pydata.org/docs/reference/api/pandas.concat.html#pandas.concat) makes a full copy of the data, and iteratively reusing [concat()](https://pandas.pydata.org/docs/reference/api/pandas.concat.html#pandas.concat) can create unnecessary copies. Collect all [DataFrame](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.html#pandas.DataFrame) or [Series](https://pandas.pydata.org/docs/reference/api/pandas.Series.html#pandas.Series) objects in a list before using [concat()](https://pandas.pydata.org/docs/reference/api/pandas.concat.html#pandas.concat)." However, for our use case I guess we can do this only once so should also be ok. 


And tested it -- seems like `concat` is the way to go in terms of performance:

```
%timeit pd.concat([df:(10000, 5000), cols:2000], axis=1)
280 ms ± 22.3 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

%timeit df:(10000, 5000).assign(**cols:2000)
4.44 s ± 765 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

%memit pd.concat([df:(10000, 5000), cols:2000], axis=1)
peak memory: 2107.36 MiB, increment: 0.20 MiB

%memit df:(10000, 5000).assign(**cols:2000)
peak memory: 2525.92 MiB, increment: 37.08 MiB
```

```python
%load_ext memory_profiler
import numpy as np
import pandas as pd
import warnings  
warnings.simplefilter(action='ignore', category=pd.errors.PerformanceWarning)

pd_size = 10000
df = pd.DataFrame(np.random.randint(0,pd_size,size=(pd_size, pd_size//2)))#, columns=list('ABCD'))

cols = []
for i in range(pd_size//10):
    cols.append(np.random.rand(pd_size))

concat_cols = [pd.Series(col) for col in cols]
assign_cols = {f"{pd_size+i+1}":col for i,col in enumerate(cols)}


print(f'\n%timeit pd.concat([df:{df.shape}, cols:{len(concat_cols)}], axis=1)')
temp_df = [df.copy()]
temp_df.extend(concat_cols)
%timeit l=pd.concat(temp_df, axis=1)

print(f'\n%timeit df:{df.shape}.assign(**cols:{len(assign_cols)})')
temp_df = df.copy()
%timeit temp_df.assign(**assign_cols)


print(f'\n%memit pd.concat([df:{df.shape}, cols:{len(concat_cols)}], axis=1)')
temp_df = [df.copy()]
temp_df.extend(concat_cols)
%memit l=pd.concat(temp_df, axis=1)

print(f'\n%memit df:{df.shape}.assign(**cols:{len(assign_cols)})')
temp_df = df.copy()
%memit temp_df.assign(**assign_cols)
```